### PR TITLE
Update comment on the comment form

### PR DESCRIPTION
**Pull request by joaojusto**
Open comment for edition on the comment form.

### CODE REVIEW

File: app/assets/javascripts/app/components/comment_form/comment_form.js [hunk 1]
```diff
@@ -14,22 +15,37 @@
...
+      $scope.comment.body = '';
+    }
+
+    function editComment() {
+
+    }
+
+    function createComment(comment) {
+      $scope.discussion
+      .addComment(comment)
```
Comment (zamith) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: Is this the style we've adopted? Shouldn't this be indented? @gabrielpoca @ronaldofs @azevedo-252 @naps62

Comment (gabrielpoca) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: We have not specified, but if it's in two lines then it must be indented.

Comment (zamith) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: Actually, we have it specified: https://github.com/groupbuddies/guides/tree/master/style#formatting

> Indent continued lines two spaces.

Comment (joaojusto) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: Ok

Comment (gabrielpoca) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: @zamith I don't recall participating in any discussion on that.

Comment (zamith) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: This is the styling for all the languages we use. There was no discussion for this in particular, it has been there since we added the guides. If you have any issue with it, open a pull request and it can be discussed there.

File: app/assets/javascripts/app/components/comment_form/comment_form.js [hunk 2]
```diff
@@ -3,7 +3,8 @@
 
   angular
     .module('council.components')
-    .directive('commentForm', commentForm);
+    .directive('commentForm', commentForm)
+    .controller('CommentFormCtrl', CommentFormCtrl);
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: Please one entity (directive, controller, ...) per file

File: app/assets/javascripts/app/components/comment_form/comment_form.js (hunk 1 above)
Comment (gabrielpoca) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: Nop, I agree with that.

File: app/assets/javascripts/app/components/comment_form/comment_form.js [hunk 3]
```diff
@@ -14,22 +15,33 @@
...
-        };
+    ctrl.clearBody = clearBody;
+    ctrl.updateComment = updateComment;
+    ctrl.createComment = createComment;
 
-        $scope.create = createComment;
-      }
-    };
+    function clearBody() {
+      $scope.comment.body = '';
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: You can assign isolate scope's properties directly to the controller by adding ```bindToController: true``` to the *DDO*.
This means you can stop using `$scope`.

```
function commentForm() {
    return {
      ...,
      bindToController: true  
    };
  }

function CommentFormCtrl(DS) {
    ...
    function clearBody() {
      ctrl.comment.body = '';
    }

    function createComment(comment) {
     ctrl.discussion
        .addComment(comment)
        .then(clearBody);
      ctrl.toggle();
    }

    function updateComment(comment) {
      var updated = DS.createInstance('discussionComment', comment, { useClass:true });
      updated.update(comment.body);
      ctrl.toggle();
    }
  }
```

File: app/assets/javascripts/app/components/comment_form/comment_form.js [hunk 4]
```diff
@@ -14,22 +15,33 @@
...
+    }
+
+    function createComment(comment) {
+      $scope.discussion
+        .addComment(comment)
+        .then(clearBody);
+      $scope.toggle();
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: Isn't this a string? How come you're calling it as a function? Am i missing something?

File: app/assets/javascripts/app/components/comment_form/comment_form.js [hunk 5]
```diff
@@ -14,22 +15,33 @@
...
+    }
+
+    function updateComment(comment) {
+      var updated = DS.createInstance('discussionComment', comment, { useClass:true });
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: I hope this gets updated for better with @gabrielpoca's new branch. This seems very ad hoc

File: app/assets/javascripts/app/discussions/show.html.slim [hunk 6]
```diff
@@ -30,7 +30,7 @@
             span am-time-ago="ctrl.discussion.created_at"
 
       md-content.md-padding md-theme="white"
-        div comments-list="ctrl.discussion.comments" comment-count="ctrl.discussion.comments_count"
+        div comments-list="ctrl.discussion.comments" comment-count="ctrl.discussion.comments_count" toggle="ctrl.toggleEditComment"
 
 sliding-screen show="ctrl.showNewComment" hide-bottom=""
-  comment-form toggle="ctrl.toggleNewComment" comment="ctrl.newComment" discussion="ctrl.discussion"
+  comment-form toggle="ctrl.toggleNewComment" comment="ctrl.newComment" discussion="ctrl.discussion" edit="ctrl.updateComment"
```
Comment (azevedo-252) [inline app/assets/javascripts/app/discussions/show.html.slim]: I can't find that ```edit``` attribute in the directive's definition

File: app/assets/javascripts/app/components/comment_form/comment_form.js (hunk 2 above)
Comment (joaojusto) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: Ok

File: app/assets/javascripts/app/components/comment_form/comment_form.js (hunk 4 above)
Comment (joaojusto) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: No this is a function

Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/comment_form.js]: Ok it was my bad

File: app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
+(function() {
+
+  angular
+  .module('council.components')
+  .controller('CommentFormCtrl', CommentFormCtrl);
+
+  function CommentFormCtrl(DS) {
+    var ctrl = this;
+
+    ctrl.resetComment = resetComment;
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js]: This shouldn't be publicly exposed if not needed

File: app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    ctrl.updateComment = updateComment;
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js]: I guess having the suffix comment in this controller (which belongs to the comment directive and is used in the comment view, basically a comment component) is kind of redundant

File: app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim [hunk 9]
```diff
@@ -2,12 +2,16 @@
   .TopBarLayout-header
     md-toolbar on-scroll="" on-scroll-anchor="#js-comment-content" scrolling-class="md-whiteframe-z1"
       md-content.md-padding layout="row" layout-align="space-between center"
-        md-button.md-primary.md-warn md-no-ink="" ng-click="toggle()" layout="row" layout-align="center center"
+        md-button.md-primary.md-warn md-no-ink="" ng-click="ctrl.toggle()" layout="row" layout-align="center center"
           .IconClose
           | hide
 
-        council-raised-button.Button.Button--secondary ng-click="create(comment)"
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim]: For future reference: i believe that having all context made available to `create` method (or any controller method) via explicitly passing all required arguments instead of relying on the controller's scope is better. Mainly because it makes them more isolated, which in turn makes them easier to test

File: app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim [hunk 10]
```diff
@@ -2,12 +2,16 @@
...
+        council-raised-button.Button.Button--secondary ng-click="ctrl.createComment()" ng-hide="ctrl.comment.id"
           | Add Comment
+
+        council-raised-button.Button.Button--secondary ng-click="ctrl.updateComment()" ng-show="ctrl.comment.id"
+          | Update Comment
+
   .TopBarLayout-content
     .Page.Page--white.u-fullSize
-      textarea.Editor placeholder="State your arguments" ng-model="comment.body" id="js-comment-content"
+      textarea.Editor placeholder="State your arguments" ng-model="ctrl.comment.body" id="js-comment-content"
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim]: ```textarea#js-comment-content.Editor placeholder="State your arguments" ng-model="ctrl.comment.body"```

in case you miss what i did, you can move write the id like that in *slim*

File: app/assets/javascripts/app/components/comments_list/comments_list.js [hunk 11]
```diff
@@ -3,48 +3,16 @@
...
-    .controller('CommentsListCtrl', CommentsListCtrl);
+    .directive('commentsList', CommentsList);
 
   function CommentsList() {
     return {
       scope: {
         comments: '=commentsList',
-        count: '=commentCount'
+        count: '=commentCount',
+        toggle: '=toggle'
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comments_list/comments_list.js]: You should be consistent. This should be an expression as well.

File: app/assets/javascripts/app/components/comments_list/comments_list.js [hunk 12]
```diff
@@ -3,48 +3,16 @@
...
       },
-      controller: 'CommentsListCtrl as ctrl',
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comments_list/comments_list.js]: I'm not sure it works or not. But you do not have a controller in this directive. That shouldn't be a cue to go back to using `$scope`again.

File: app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js (hunk 7 above)
Comment (joaojusto) [inline app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js]: Ok

File: app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js (hunk 8 above)
Comment (joaojusto) [inline app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js]: You are right

File: app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim (hunk 9 above)
Comment (joaojusto) [inline app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim]: So do you think I should pass the arguments instead? I think this way the html is a lot more cleaner

File: app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim (hunk 10 above)
Comment (joaojusto) [inline app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim]: Is I know that, in this case I prefer to put it this way, no reason at all it's just personal preference. I can change that

File: app/assets/javascripts/app/components/comments_list/comments_list.js (hunk 11 above)
Comment (joaojusto) [inline app/assets/javascripts/app/components/comments_list/comments_list.js]: Ok

File: app/assets/javascripts/app/components/comments_list/comments_list.js (hunk 12 above)
Comment (joaojusto) [inline app/assets/javascripts/app/components/comments_list/comments_list.js]: Ok

**Comment by joaojusto** [comment]
@azevedo-252 bump

**Comment by azevedo-252** [comment]
@gabrielpoca merged his branch. Now please rebase this branch and make appropriate changes

**Comment by joaojusto** [comment]
Ok

File: app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+    ctrl.update = update;
+    ctrl.create = create;
+
+    function reset() {
+      ctrl.comment = Comment.createInstance();
+    }
+
+    function create() {
+      ctrl.discussion
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js]: I've been thinking that maybe it makes more sense for a directive's controller to be in the same file as the directive. Even more so if it has `bindToController: true`.

These kind of properties that come from the directive's isolate scope, always make you go back to the directive's file to know what it is.
@ronaldofs @gabrielpoca What do you think?

File: app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim [hunk 14]
```diff
@@ -1,12 +1,16 @@
 .TopBarLayout
   .TopBarLayout-header
-    md-toolbar md-theme="white" on-scroll="" on-scroll-anchor="#js-comment-content" scrolling-class="md-whiteframe-z1" layout="row" layout-align="space-between center"
-        md-button.md-primary.md-warn md-no-ink="" ng-click="toggle()" md-theme="council" layout="row" layout-align="space-between center"
-          .IconClose
-          | hide
+    md-toolbar on-scroll="" md-theme="white" on-scroll-anchor="#js-comment-content" scrolling-class="md-whiteframe-z1" layout="row" layout-align="space-between center"
```
Comment (azevedo-252) [inline app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim]: why the change?

Comment (joaojusto) [inline app/assets/javascripts/app/components/comment_form/templates/comment_form.html.slim]: No resason, it was on the rebase that I changed the order

File: app/assets/javascripts/app/discussions/discussion.js [hunk 15]
```diff
@@ -29,7 +31,12 @@
       discussion.update();
     }
 
-    function toggleNewComment() {
+    function openEditForm(params) {
+      ctrl.comment = params.comment;
```
Comment (azevedo-252) [inline app/assets/javascripts/app/discussions/discussion.js]: You should be expecting the actual comment and not an object. It doesn't make much sense like this

File: app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js (hunk 13 above)
Comment (gabrielpoca) [inline app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js]: I don't know. It's on the same folder right?

File: app/assets/javascripts/app/discussions/discussion.js [hunk 16]
```diff
@@ -5,12 +5,14 @@
...
+  function DiscussionCtrl(Discussion, Comment, discussionId) {
     var ctrl = this;
 
     ctrl.pageReady = false;
-    ctrl.toggleNewComment = toggleNewComment;
-    ctrl.toogleDiscussionState = toogleDiscussionState;
+    ctrl.openEditForm = openEditForm;
+    ctrl.toggleCommentForm = toggleCommentForm;
+
+    ctrl.comment = Comment.createInstance();
```
Comment (azevedo-252) [inline app/assets/javascripts/app/discussions/discussion.js]: This instance creation should only exist in one place. Make this happen in the form controller alone

File: app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js (hunk 13 above)
Comment (joaojusto) [inline app/assets/javascripts/app/components/comment_form/comment_form_ctrl.js]: Yes they are

**Comment by joaojusto** [comment]
@azevedo-252 don't forget to review this

**Comment by azevedo-252** [comment]
:+1:
